packages/store/test/test-heap-classes.js

```diff
@@ -0,0 +1,107 @@
+// @ts-check
+
+import { test } from '@agoric/swingset-vat/tools/prepare-test-env-ava.js';
+import {
+  defineHeapFarClass,
+  defineHeapFarClassKit,
+  makeHeapFarInstance,
+} from '../src/patterns/interface-tools.js';
+import { M } from '../src/patterns/patternMatchers.js';
+
+const UpCounterI = M.interface('UpCounter', {
+  incr: M.call()
+    // TODO M.number() should not be needed to get a better error message
+    .optional(M.and(M.number(), M.gte(0)))
+    .returns(M.number()),
+});
+
+const DownCounterI = M.interface('DownCounter', {
+  decr: M.call()
+    // TODO M.number() should not be needed to get a better error message
+    .optional(M.and(M.number(), M.gte(0)))
+    .returns(M.number()),
+});
+
+test('test defineHeapFarClass', t => {
+  const makeUpCounter = defineHeapFarClass(
+    'UpCounter',
+    UpCounterI,
+    (x = 0) => ({ x }),
+    {
+      incr(y = 1) {
+        // @ts-expect-error TS doesn't know that `this` is a `Context`
```

Comment:
Will we be able to tell the type system?  That error is going to be a recurring pain point.

Reply:
I have no idea, and I agree. I did try but was helpless in the face of TypeScript's rules. 

@turadg @mhofman @michaelfig ?